core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,169 @@
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+
+        Map<ColumnHandle, Constant> assignments = buildAssignments(orderedColumnNames, mergeAssignments, columnHandles, context);
+        if (assignments.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Query Planner builds the merge rowblock, containing:
+        // All data columns in table order
+        // plus 3 additional fields: field present flag, update operation number, merge case number
+        // TODO move this check before buildAssignments as soon support for update all will be added,
+        //  in case of update all - row encoded as base64, which filtered in buildAssignments
+        checkState(orderedColumnNames.size() == mergeAssignments.size() - 3, "Number of data columns should be equal");
+
+        return metadata.applyUpdate(context.getSession(), tableScan.getTable(), assignments)
+                .map(newHandle -> new TableUpdateNode(
+                        context.getIdAllocator().getNextId(),
+                        newHandle,
+                        getOnlyElement(node.getOutputSymbols())))
+                .map(Result::ofPlanNode)
+                .orElseGet(Result::empty);
+    }
+
+    private Map<ColumnHandle, Constant> buildAssignments(
+            List<String> orderedColumnNames,
+            List<? extends Node> mergeAssignments,
+            Map<String, ColumnHandle> columnHandles,
+            Context context)
+    {
+        ImmutableMap.Builder<ColumnHandle, Constant> assignmentsBuilder = ImmutableMap.builder();
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            if (isSymbol(assigmentNode)) {
+                // the column is not updated
+                continue;
+            }
+            if (!isConstant(assigmentNode)) {
+                return ImmutableMap.of();
+            }
+
+            ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                    context.getSession(),
+                    ((Expression) assigmentNode),
+                    context.getSymbolAllocator().getTypes(),
+                    plannerContext,
+                    typeAnalyzer);
+
+            ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();
+            // we don't support any expressions in update statements yet, only constants
+            // also in case of UPDATE ALL statement - row represented as base64 call
+            if (!(connectorExpression instanceof Constant)) {
+                return ImmutableMap.of();
+            }
```

Comment:
Using `translateConjuncts` this way is incorrect. That method returns a structure that contains a translated part and a residual part. This code is ignoring the residual, which contains any parts of the expression that could not be translated. Also, `translateConjuncts` is designed to deal with boolean expressions (hence the term "conjuncts" -- the component terms of a boolean expression connected by `AND` operators).

Use `ConnectorExpressionTranslator.translate()` instead, which will return `Optional.empty()` if the expression could not be translated, or the complete translated expression otherwise.